src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/RandomValueTokenFactory.java

```diff
@@ -0,0 +1,89 @@
+package com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token;
+
+import com.atlassian.bitbucket.jenkins.internal.applink.oauth.Randomizer;
+import com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.consumer.Consumer;
+import com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.exception.InvalidTokenException;
+import net.oauth.OAuthMessage;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.net.URI;
+import java.util.logging.Logger;
+
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token.ServiceProviderToken.newAccessToken;
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token.ServiceProviderToken.newRequestToken;
+import static java.lang.System.currentTimeMillis;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+/**
+ * Implementation of {@link TokenFactory} that uses randomly generated values for tokens and delegates storage of
+ * generated tokens to {@link ServiceProviderTokenStore}.
+ * <br>
+ * The random token generation is based on the
+ * <a href="https://github.com/spring-projects/spring-security-oauth/blob/master/spring-security-oauth/src/main/java/org/springframework/security/oauth/provider/token/RandomValueProviderTokenServices.java">Spring OAuth Provider Library</a>
+ */
+@Singleton
+public class RandomValueTokenFactory implements TokenFactory {
```

Comment:
I would rename the class to something which is more meaningful. Random in the name is sort of leaking unimportant detail which client of the class need not know. How about `ServiceProviderTokenFactory`?

Reply:
The name was stolen from the Spring oauth lib ;-)
`TokenFactory` vs. `ServiceProviderTokenFactory` IMO implies that there can be a non-ServiceProvider `TokenFactory` implementation :thinking: 
How about some lazy naming like `DefaultTokenFactory` or something like that?